### ISSUE TITLE
Improve regex capabilities or update documentation

### ISSUE THREAD

**Issue by timelessvirtues** [lachesis/scallion#69]
I'm trying to apply more advanced regex patterns without luck.  I would like to run something like:

`([2-9]?)(hello)(there?)([2-9]?)|([2-9]?)(pattern)(s?)([2-9]?)|([2-9]?)(abc(x|zy))`

Is this a regex dialect issue, a limitation of converting the mask for use on the GPU or something else entirely?

**Comment by lachesis** [comment]
The "regex" is extremely limited. We currently support `.`, anchors (`$` and `^`), explicitly specified character classes (e.g. `[abcde]` but not `[a-e]`), plus the shorthand operators `\d`, `\D`, and `\w`. There's no support for groups, optional values, or even repetition operators.

This was basically created as a very limited improvement over just allowing several fixed patterns to be specified. We've actually found that the speed of scallion is very dependent on the complexity of the pattern matching code, so we chose to go with a really simple fixed size hash table lookup on the GPU and do all the complexity on the CPU. As such, we didn't (and really couldn't) implement a full regex engine.

I agree that we should ideally update the documentation. I'd be happy to merge a PR that updates the README.
